pkg/azure/provider/provider_test.go

```diff
@@ -0,0 +1,1180 @@
+		vmAccessAPIBehaviorSpec   *fakes.APIBehaviorSpec
+		rgAccessAPIBehaviorSpec   *fakes.APIBehaviorSpec
+		diskAccessAPIBehaviorSpec *fakes.APIBehaviorSpec
+		nicAccessAPIBehaviorSpec  *fakes.APIBehaviorSpec
+		cascadeDeleteOpts         fakes.CascadeDeleteOpts
+		vmExists                  bool
+		checkErrorFn              func(g *WithT, err error)
+		checkClusterStateFn       func(g *WithT, ctx context.Context, clusterState *fakes.ClusterState, vmName string)
+	}{
+		{
+			"should fail when checking resource groups existence returns an error", nil,
+			fakes.NewAPIBehaviorSpec().AddErrorResourceReaction(testResourceGroupName, testhelp.AccessMethodCheckExistence, testInternalServerError),
+			nil, nil, fakes.CascadeDeleteAllResources, true, checkError,
+			func(g *WithT, ctx context.Context, clusterState *fakes.ClusterState, vmName string) {
+				createFactoryAndCheckClusterState(g, ctx, testResourceGroupName, clusterState, vmName, true, true, true)
+			},
+		},
+		{
+			"should fail when VM access Get call returns an error",
+			fakes.NewAPIBehaviorSpec().AddErrorResourceReaction(vmName, testhelp.AccessMethodGet, testInternalServerError),
+			nil, nil, nil, fakes.CascadeDeleteAllResources, true, checkError,
+			func(g *WithT, ctx context.Context, clusterState *fakes.ClusterState, vmName string) {
+				createFactoryAndCheckClusterState(g, ctx, testResourceGroupName, clusterState, vmName, true, true, true)
+			},
+		},
+		{
+			"non-existing-vm: should delete left over OSDisk even if error is returned when deleting left over NIC",
+			nil, nil, nil,
+			fakes.NewAPIBehaviorSpec().AddErrorResourceReaction(utils.CreateNICName(vmName), testhelp.AccessMethodBeginDelete, testInternalServerError),
+			fakes.CascadeDeleteAllResources, false, checkError,
+			func(g *WithT, ctx context.Context, clusterState *fakes.ClusterState, vmName string) {
+				createFactoryAndCheckClusterState(g, ctx, testResourceGroupName, clusterState, vmName, false, true, false)
+			},
+		},
+		{
+			"non-existing-vm: should delete left over NIC even if there is a panic when deleting left over OSDisk",
+			nil, nil,
+			fakes.NewAPIBehaviorSpec().AddPanicResourceReaction(utils.CreateOSDiskName(vmName), testhelp.AccessMethodBeginDelete),
+			nil, fakes.CascadeDeleteAllResources, false, checkError,
+			func(g *WithT, ctx context.Context, clusterState *fakes.ClusterState, vmName string) {
+				createFactoryAndCheckClusterState(g, ctx, testResourceGroupName, clusterState, vmName, false, false, true)
+			},
+		},
+		{
+			"should fail when existing VM's cascade delete options update returns an error",
+			fakes.NewAPIBehaviorSpec().AddErrorResourceReaction(vmName, testhelp.AccessMethodBeginUpdate, testInternalServerError),
+			nil, nil, nil, fakes.CascadeDeleteOpts{}, true, checkError,
+			func(g *WithT, ctx context.Context, clusterState *fakes.ClusterState, vmName string) {
+				createFactoryAndCheckClusterState(g, ctx, testResourceGroupName, clusterState, vmName, true, true, true)
```

Comment:
here also we should check that cascadeDeleteOptions is not set for the machineResources using `checkCascadeDeleteOptions()`

Reply:
Ok added that check